### PR TITLE
CIP-0073? | Oracle Datum Standard

### PR DESCRIPTION
**Pull request by codybutz**
This CIP proposes a standard for Oracle Datum formats. The goal of this initial CIP is to provide the standard for KYC and Price feeds. However, this Oracle CIP is intended to be a living standard, allowing other forms of Datums to be available to be standardized. For example, this CIP could be extended to include standards for weather data, population data, etc.

Charli3 is finalizing a open-source PlutusTX library for this standard.

### CODE REVIEW

**Comment by peterVG** [comment]
This CIP is of obvious interest to the Orcfax oracle project as well. We support the rationale for an oracle datum standard and would like to make it compatible within our Cardano Open Oracle Protocol which is a slightly broader, opinionated design for requesting and consuming oracle datum on Cardano. We have just learned of this draft update and will need time to schedule our own review before we can contribute any meaningful input if that is useful/desirable.

File: CIP-XXXX/standard.cddl [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+; A price_data entry normally defines the “Currency Pair” being priced.
+;  A “Currency Pair” is defined as “the first listed currency or base currency is
+;  bought while the second listed currency in a currency pair or quote currency is
+;  sold.” Each currency is defined separately in the entry, and referred to as the
+;  “base” and “quote” currency.
+price_data = #6.123([price_map])
+price_map =
+  { ? 0 : price        ; how many quote currency is received per base currency spent
+  , ? 1 : posixtime    ; unix timestamp related to when the price data was created
+  , ? 2 : posixtime    ; unix timestamp related to when the price data is expired
```
Comment (bladyjoker) [inline CIP-XXXX/standard.cddl:32]: Considering using a single time to denote when the value was 'observed'. In frp terminology, price is a function over 'continuous' time and it simplifies working with it in such way. So one could just have a field `? 1 : posixtime ; unix timestamp of when the price was observed`.

Any script running onchain can make their own decision on how much 'interval' to put around an observation by simply saying `contains someIntervalAroundObservation observedAt`

File: CIP-XXXX/standard.cddl [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+; The extended data carries provider-specific information, and may
+;  be individually customized by oracle providers to provide data
+;  they deem relevant. It's thus not entirely defined by spec,
+;  but indices below 100 are "reserved" by the spec.
+extended_data = #6.122(extended_map)
+extended_map = 
+  { ? 0: oracle_provider_id
+  , ? 1: data_source_count
+  , ? 2: data_signatories_count
+  , ? 3: oracle_provider_signature
```
Comment (bladyjoker) [inline CIP-XXXX/standard.cddl:66]: Not sure what this signature will represent? The oracle signs some X, what is X?

**Comment by bladyjoker** [comment]
Hey @codybutz nice work, I'm glad that structured data representation is being discussed as it directly relates to the project we're running funded by Catalyst (https://cardano.ideascale.com/c/idea/421376). 

Allow me to paint a different picture of how this could work, that wouldn't require any adoption at the Cardano level via a CIP and CBOR/CDDL wouldn't even have to be mentioned. It would all rely on standardized 'PlutusData' encoding for any type specified in a configuration file. 

Perhaps it's useful to start with listing related technologies in this space and the engineering principles they embody:
1. Google Protobuf
2. Apache Thrift
3. ASN.1
4. And indeed CDDL

These technologies enable a very important separation of concerns between what's called an 'Abstract syntax' (ie. your types) and 'Concrete syntax' (ie. values of your types, encodings).
CDDL is of course one such technology, but it is in many ways inappropriate due to it's complexity and very limited tooling support, but also it is highly coupled with its 'Concrete syntax' which is CBOR.

When specifying 'shared types' we must consider a broader perspective in which these types are going to be shared in. For instance, backend-frontend communication, storage, databases and of course when building Cardano transactions. Each of these settings if not accounted for will require a lot of manual effort, and this proposal imo could have nicer gains if a different route altogether is considered.

In our project we'd aim to support definitions of types much like you're used to do in Haskell...

oracle.lambuff
```haskell
-- | NOTE: Not Haskell! But an IDL language inspired by Haskell
module Cardano.Oracle where

import Plutus (AssetClass, POSIXTime)

-- | OracleDatum is a generic Oracle envelope
data OracleDatum a = OracleDatum {
  provenance :: OracleProvenance
  datum :: a
}

-- | Provenance information provided with each datum
data OracleProvenance = OracleProvenance {
  providerId :: Nat,
  dataSourceCount :: Nat,
  dataSignatoriesCount :: Nat,
  oracleSignature :: Bytes
}

data Price = Price {
  ratio :: Decimal,
  at :: POSIXTime,
  baseAsset :: AssetClass,
  quoteAsset  :: AssetClass
}

type PriceOracleDatum = OracleDatum Price
```

Now specifying types is ofc only one part of the story, the second part is where actual value comes in:
1. Codegen Purescript and Haskell type libraries with runtime support for standardized PlutusData and JSON encodings
2. Codegen tailored for common PAB backends such as cardano-transaction-lib, and common Plutus frontends such as PlutusTx and Plutarch
3. Backwards compatibility management
4. Specified and thoroughly tested cross-language compatibility

What this would mean in short is that Charli3 and Orcfax could agree on a 'Oracle schema' they publish as a configuration file in a repository and would have to provide zero code. After that users can leverage our tooling to have these types communicable in any of the aforementioned scenarios (transaction building, plutus programming, databases and storage, analytics, frontend-backend communication).

Apologies for the huge rant, I just wanted to take the opportunity to advertise our project and illustrate its goals.

**Review by GeorgeFlerovsky — COMMENTED** [review]
I agree with the motivations of this CIP, and think that oracle projects should do more over the coming year to improve the experience for data consuming projects to consume multi-provider oracle data.

My main objection to the CIP is that using integer CBOR tags inevitably leads to global/central registries, which could become centralised gatekeepers. There are alternative approaches to coordinating the consumer-provider interaction that would not lead to this centralisation -- in particular, tagging content with a schema hash would avoid the need to look at a global registry. (See my specific review comment)

More narrowly, this CIP standard is unfortunately not compatible with the datum structure used by Cardano Open Oracle Protocol (COOP) and Orcfax. In the medium-longer term, I'm sure we can find a way to reconcile them, but it's not feasible in the short term (i.e. Indigo launch / COOP release).

In general, I think that this CIP may be too ambitious right now. It would probably be more beneficial to have more domain-specific standards for several particular usecases, and then see the emerging unifying theme / best practices emerge over the course of the following year.

For now, I would recommend that Indigo proceed with what it's already committed to, and perhaps even to simplify the CIP to remove the generalisation features, as I suspect this and other standards will be rather transient over the course of the coming year. Particularly, I would hold off on launching the global Cardano tag registry and oracle provider registry.

File: CIP-XXXX/README.md [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+   <td>Integer (boolean)
+   </td>
+   <td>Address is whitelisted for sale
+   </td>
+  </tr>
+</table>
+
+## Rationale
+
+This CIP is ready to be made active; since it is a process standard, it requires no implementation.
```
Comment (KtorZ) [inline CIP-XXXX/README.md:308]: Actually, it does require projects adopting the standard. This is usually a good requirement for the **Path to Active.** section: have several projects adopting the standard specified in the document. If there are already some known projects that intend to, they can also explicitly be listed here. 

Note also that this covers the **Path to Active** section, but not the **Rationale** section. Rationale is about providing justifications for the various design decisions made in the proposed document. For example, it can show how the solution compares to the state of the art on other blockchain ecosystem, or explain the rationale behind the choice of a particular technology. The rationale may also explain why some solutions that could have been considered were rejected.

File: CIP-XXXX/README.md [hunk 4]
```diff
@@ -0,0 +1,312 @@
+---
+CIP: <?>
```
Comment (KtorZ) [inline CIP-XXXX/README.md:2]:
```suggestion
CIP: 73  
```

File: CIP-XXXX/README.md [hunk 5]
```diff
@@ -0,0 +1,312 @@
+---
+CIP: <?>  
+Title: <Oracle Datum Standard>
```
Comment (KtorZ) [inline CIP-XXXX/README.md:3]:
```suggestion
Title: Oracle Datum Standard
```

File: CIP-XXXX/README.md [hunk 6]
```diff
@@ -0,0 +1,312 @@
+---
+CIP: <?>  
+Title: <Oracle Datum Standard>  
+Authors: Cardano DeFi Alliance <info@cardanodefialliance.org>, Cody Butz <cbutz@indigo-labs.io>, Jonas Lindgren <jonas@charli3.io>
+Comments-Summary: <No comments>  
+Comments-URI: <>  
+Status: <Draft>
```
Comment (KtorZ) [inline CIP-XXXX/README.md:7]:
```suggestion
Status: Proposed
```

File: CIP-XXXX/README.md [hunk 7]
```diff
@@ -0,0 +1,312 @@
+---
+CIP: <?>  
+Title: <Oracle Datum Standard>  
+Authors: Cardano DeFi Alliance <info@cardanodefialliance.org>, Cody Butz <cbutz@indigo-labs.io>, Jonas Lindgren <jonas@charli3.io>
+Comments-Summary: <No comments>  
+Comments-URI: <>  
+Status: <Draft>  
+Type: <Standards>
```
Comment (KtorZ) [inline CIP-XXXX/README.md:8]:
```suggestion
Type: Process  
```

File: CIP-XXXX/README.md [hunk 8]
```diff
@@ -0,0 +1,312 @@
+---
+CIP: <?>  
+Title: <Oracle Datum Standard>  
+Authors: Cardano DeFi Alliance <info@cardanodefialliance.org>, Cody Butz <cbutz@indigo-labs.io>, Jonas Lindgren <jonas@charli3.io>
+Comments-Summary: <No comments>  
+Comments-URI: <>  
+Status: <Draft>  
+Type: <Standards>  
+Created: <2022-10-20>
```
Comment (KtorZ) [inline CIP-XXXX/README.md:9]:
```suggestion
Created: 2022-10-20
```

File: CIP-XXXX/README.md [hunk 9]
```diff
@@ -0,0 +1,312 @@
+---
+CIP: <?>  
+Title: <Oracle Datum Standard>  
+Authors: Cardano DeFi Alliance <info@cardanodefialliance.org>, Cody Butz <cbutz@indigo-labs.io>, Jonas Lindgren <jonas@charli3.io>
+Comments-Summary: <No comments>  
+Comments-URI: <>  
+Status: <Draft>  
+Type: <Standards>  
+Created: <2022-10-20>  
+License: <CC-BY-4.0>
```
Comment (KtorZ) [inline CIP-XXXX/README.md:10]:
```suggestion
License: CC-BY-4.0
```

File: CIP-XXXX/README.md [hunk 10]
```diff
@@ -0,0 +1,312 @@
+---
+CIP: <?>  
+Title: <Oracle Datum Standard>  
+Authors: Cardano DeFi Alliance <info@cardanodefialliance.org>, Cody Butz <cbutz@indigo-labs.io>, Jonas Lindgren <jonas@charli3.io>
+Comments-Summary: <No comments>  
+Comments-URI: <>
```
Comment (KtorZ) [inline CIP-XXXX/README.md:6]:
```suggestion
```

File: CIP-XXXX/README.md [hunk 11]
```diff
@@ -0,0 +1,312 @@
+---
+CIP: <?>  
+Title: <Oracle Datum Standard>  
+Authors: Cardano DeFi Alliance <info@cardanodefialliance.org>, Cody Butz <cbutz@indigo-labs.io>, Jonas Lindgren <jonas@charli3.io>
+Comments-Summary: <No comments>
```
Comment (KtorZ) [inline CIP-XXXX/README.md:5]:
```suggestion
```

File: CIP-XXXX/standard.cddl [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+  , ? 1: kyc_data
+  }
+
+; The extended data carries provider-specific information, and may
+;  be individually customized by oracle providers to provide data
+;  they deem relevant. It's thus not entirely defined by spec,
+;  but indices below 100 are "reserved" by the spec.
+extended_data = #6.122(extended_map)
+extended_map = 
+  { ? 0: oracle_provider_id
```
[the remark this answers is not in the document]
Comment (ipo) [reply CIP-XXXX/standard.cddl:63]: If there is not a registry of which ids represent which providers, but dApp consumers are "fine to whitelist the providers", how would the dApp consumers know which ids represent which providers, in such a way that collisions aren't happening, without a registry?

I think you are reading too much "control" into a registry, it's not for a registry to accept and deny, it's for a registry to keep obvious spam out and prevent collisions.

File: CIP-XXXX/README.md [hunk 13]
```diff
@@ -0,0 +1,312 @@
...
+#### Detailed Format
+
+Any data feed datum matching this standard must adhere to the following overall structure. Each of these individual elements is a Tag defined further later on.
+
+The data structure consists of a list of Tags. This list is itself surrounded by a tag versioned to this major version of CIP. A later CIP may thus build on top of this standard by adding further datum after the initial one, using a different tag that allows older users of this CIP to disregard this new data.
+
+There are 3 major categories of Tags, described in further detail later on, SharedData, GenericData and ExtendedData. In the list the ordering is defined as [ ? SharedData, 1* GenericData, ? ExtendedData ]. Or in plain english, a SharedData may optionally come first, then 1 or more GenericData in no particular order, and finally an ExtendedData may optionally come last.
+
+
+### CIP CBOR Tags Registry
```
[the remark this answers is not in the document]
Comment (ipo) [reply CIP-XXXX/README.md:58]: Orcfax were invited to participate in the process, we were very open, but there didn't seem to be a great deal of interest at the time.

File: CIP-XXXX/standard.cddl [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+  , ? 1: data_source_count
+  , ? 2: data_signatories_count
```
[the remark this answers is not in the document]
Comment (ipo) [reply CIP-XXXX/standard.cddl:65]: As per the specification, the part you are now criticizing is `individually customized by oracle providers to provide data they deem relevant`, so what is generally applicable to everyone doesn't matter here. This is the part where each data provider would be putting things specific to their implementation, that does not fit into a general, shared standard.

File: CIP-XXXX/README.md (hunk 3 above)
Comment (ipo) [reply CIP-XXXX/README.md:308]: I think that is a fair expectation. As has been mentioned before Charli3 are currently implementing this standard in publicly accessible, partially Catalyst-funded community feeds, which we hope can help wider adoption of the standard.

File: CIP-XXXX/README.md (hunk 7 above)
Comment (ipo) [reply CIP-XXXX/README.md:8]: I'm confused by the suggestion to change from Standards to Process, isn't this very clearly in the Standards camp? Am I misunderstanding the definition of Standards vs Process, or have I missed context?

Comment (ipo) [reply CIP-XXXX/README.md]: Will do, thank you.

Comment (rphair) [reply CIP-XXXX/README.md]: Thanks for edits... going back now & resolving conversations relating to preamble format.

**Comment by rphair** [comment]
@ipo thanks for edits... resolved conversations relating to preamble format.

File: CIP-XXXX/standard.cddl (hunk 12 above)
[the remark this answers is not in the document]
Comment (GeorgeFlerovsky) [reply CIP-XXXX/standard.cddl]: Suppose that there is sport betting protocol and an oracle provider for horse-racing results. The sport betting protocol wants to whitelist the oracle provider.

Under your proposal, the oracle provider would have to apply to the Cardano-wide oracle registry to obtain an integer ID, so that the sport betting protocol could whitelist that oracle provider using its integer ID.

---

Alternatively, if the oracle were instead identified by either an oracle signing key or minting policy ID, then the sport betting protocol could whitelist it simply by referring to that signing key or minting policy ID, without waiting for that oracle to be added to some registry.

A registry could still be useful in that case for dApps to discover potential oracle providers that it could whitelist, but dApps could also whitelist any oracle providers outside the registry if they so choose.

File: CIP-XXXX/README.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (GeorgeFlerovsky) [reply CIP-XXXX/README.md]: I'm not a representative of Orcfax. As an MLabs dev/manager, I collaborated with Orcfax on a Catalyst-funded project to implement the [Cardano Open Oracle Protocol (COOP)](https://www.orcfax.link/introducing-the-cardano-open-oracle-protocol-coop/).

It is unfortunate that there weren't more active discussion earlier in your CIP process to potentially converge on a single standard. My comment here merely states the fact that your CIP is incompatible with COOP because you identify oracle providers via integer ID pointers to an external registry, whereas COOP identifies oracle providers via the minting policy ID of the tokens the oracle uses to authenticate its published data. The two approaches are fundamentally not compatible (unless you allow oracle providers to be identified by either an integer ID or a minting policy ID).

[the remark this answers is not in the document]
Comment (GeorgeFlerovsky) [reply CIP-XXXX/README.md]: The two approaches can certainly co-exist in the Cardano ecosystem, and I'm curious to see which one will be more beneficial in practice.

File: CIP-XXXX/README.md [hunk 15]
```diff
@@ -0,0 +1,312 @@
...
+
+### Data Format
+
+The basis for this standard is the Concise Binary Object Representation (CBOR) data format, for which more information can be found here: [https://www.rfc-editor.org/rfc/rfc8949.html](https://www.rfc-editor.org/rfc/rfc8949.html) 
+
+A central aspect of the CBOR standard that is used extensively is the concept of a Tag, which allows data consumers to identify and consume the types of data that they expect, and gracefully ignore other pieces of data that they are not interested in.
+
+Similar to the CBOR standard this CIP will also propose a CBOR Tags Registry. Data providers are able to provide data according to the formats provided in this registry, and data consumers can discern the different available formats to pick the exact kind of data they need out of a feed.
+
+To maintain compatibility with the main CBOR Tags Registry, the tags proposed in this are defined as an offset from Base Oracle Datum CIP Tag Number, which is defined later in this document. For example, Tag +5 would correspond to the actual number OracleDatumCipTagNumber + 5.
```
[the remark this answers is not in the document]
Comment (GeorgeFlerovsky) [reply CIP-XXXX/README.md]: No, I'm not talking about transaction metadata here. I'm talking about adding something like a "_data_schema_hash_" field to the data object that gets put into the oracle utxo datum, which identifies the schema that corresponds to the data object.

The consumer of the data object is able to properly interpret the data object as long as he is aware of a data schema definition that hashes to that data schema hash. The consumer can obtain such data schema definitions from any convenient source, and has full confidence that a data schema definition obtained from any source will correspond to the data object of interest, as long as the hashes match.

File: CIP-XXXX/standard.cddl (hunk 14 above)
[the remark this answers is not in the document]
Comment (GeorgeFlerovsky) [reply CIP-XXXX/standard.cddl]: Oh, I see. So that part of your CDDL is just an example of how an oracle _could_ specify the relevant properties of how its consensus was reached on the data point?

I think my point still stands that the particular example (data source count and data signatories count) is insufficient, but it's not a blocker for your CIP if your CIP isn't mandating anything specific for oracle providers to put in there.

File: CIP-XXXX/README.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (GeorgeFlerovsky) [reply CIP-XXXX/README.md]: Please believe that I am making my comments in good faith and hoping that they are constructive for you. :pray:

File: CIP-XXXX/standard.cddl (hunk 14 above)
[the remark this answers is not in the document]
Comment (ipo) [reply CIP-XXXX/standard.cddl]: Yes, that's correct. Our expectation would be that each oracle provider would be able to (through some reasonable process) simply declare "this is our custom properties" and it would be basically merged outright.

It would probably be subject to a low bar such as "being used in practice somewhere" just to keep things relevant, but the idea is to give each provider free reign here.

The underlying idea is that this would allow oracle providers to do their own thing in areas where they disagree with the standard, while still being able to conform to the standard in the areas they do agree. Like a safety release valve for disagreements around the standard, so that implementers are minimally constrained while still allowing maximum format sharing where consensus exist.

File: CIP-XXXX/README.md [hunk 16]
```diff
@@ -0,0 +1,312 @@
...
+
+There are 3 major categories of Tags, described in further detail later on, SharedData, GenericData and ExtendedData. In the list the ordering is defined as [ ? SharedData, 1* GenericData, ? ExtendedData ]. Or in plain english, a SharedData may optionally come first, then 1 or more GenericData in no particular order, and finally an ExtendedData may optionally come last.
+
+
+### CIP CBOR Tags Registry
+
+As additional use cases and data fields become necessary, this Tags Registry should be amended and updated to add new Tags and fields inside specific tags. Care should be taken to not change or remove existing Tags and fields inside specific tags, so as to maintain backwards compatibility to as great an extent as possible.
+
+Tags are optional; data providers may provide multiple different tags covering the same data. The standard permits oracle feed providers to be flexible about what data they provide, while adhering to the standard. This allows feed consumers to programmatically discover what information is on the feed, while ignoring information that is not relevant to their use case.
```
[the remark this answers is not in the document]
Comment (ipo) [reply CIP-XXXX/README.md]: Such a tag registry could probably created in a decentralized way as well, perhaps governed by an on-chain DAO or some sort of staking-to-add-entry or even outright paying a nominal ADA amount to add entries, the amount which could then be forwarded to a charity or something to prevent a profit motive.

I don't disagree that it's centralized, but I'm not sure the benefit of decentralizing this aspect would be worth the squeeze. Additionally I think that if someone tried to unfairly assert control through the tag registry, the open nature of the standard would allow it to be forked under more democratically-minded leadership, ultimately.

File: CIP-XXXX/standard.cddl (hunk 12 above)
[the remark this answers is not in the document]
Comment (ipo) [reply CIP-XXXX/standard.cddl]: If this registry was decentralized to an on-chain registry, as I touched on in a previous comment, this issue could perhaps be avoided. The sports betting protocol could then simply reserve their ID in advance, go ahead and implement without asking for permission while at the same time dealing with a pull-request to the standard

That way, if there was a stall in the CIP process the providers and any consumers could still go ahead and use their ID with full confidence that it will be 'theirs'. The standards process might still leave the final standardization for that ID being different than what the premature implementations used, but that's just a problem inherent to any consensus-driven standards process and not really related to the ID registry.

File: CIP-XXXX/README.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (ipo) [reply CIP-XXXX/README.md]: It seems what you're proposing here is an entirely different approach, not sharing very much if anything with the proposal given here. While that approach may have advantages and disadvantages, I'm not sure this proposal is the correct venue to get into that discussion. Is there a CIP proposal or forum write-up of this somewhere where this discussion could be had?

File: CIP-XXXX/README.md [hunk 17]
```diff
@@ -0,0 +1,312 @@
...
+## Specification
+
+Oracles are responsible for providing multiple types of off-chain data to the blockchain. This proposal outlines an on-chain datum standard for Oracle producers to submit to the blockchain and Oracle consumers to read from the blockchain in a standardized format.
+
+Such a standard should accomplish a number of somewhat contradictory objectives
+
+* Data provided by the feed should be possible to alter over time, to add more detailed data or remove parts of the data, without major disruption to existing users of the data feed
+* Data provided in the feed should be stored in an efficient manner, to avoid data duplication and excessive transaction costs
+* The standard should be flexible enough to capture the majority of use-cases, even those that may not yet be conceived of at the writing of this document
+* The standard should provide room for individual data providers to add their own data fields to a feed, without interfering with the common standard
```
[the remark this answers is not in the document]
Comment (ipo) [reply CIP-XXXX/README.md]: We will always know more in the future than what we do now about how the standards are shaping up, but Charli3 (which I represent, in the interest of full disclosure) is already running free public feeds since several months back. It is in the Cardano community's interest that these feeds are provided in a broadly-useful and as standardized a way as possible, which is why we have had this discussion around standard for a long time prior to submitting this proposal, and are now requesting input from everyone else as to what is good and what needs improvement.

The sooner we begin this process, the sooner we can start to move towards something that is good. I don't imagine this standard will be perfect from the first draft and I think we should all be open to it evolving over time. Ideally in such a way that it remains backwards-compatible, that would be my biggest wish for the future.

**Comment by rphair** [comment]
@codybutz is this ever going to be submitted for review?  If you're not pursuing it anymore, would any of the other reviewers (I assume all potential advocates) be interested it taking it over, or at least posting whether this proposal was ever implemented to anyone's knowledge?

cc @peterVG @bladyjoker @GeorgeFlerovsky @klarkc

I'm marking it as `Abandoned` due to the nearly 2 years that have gone by without updates & it will be closed as a matter of course if no further progress is made: but would nice to see this move ahead as a CIP if the industry has been doing things this way.

**Comment by peterVG** [comment]
Thanks for checking in Robert. I think its safe to say that this discussion
has moved on to the BMO Intersect working group on Oracles. Cody was at one
of the first calls where he contextualized this work as an early start.
We're hoping that we can collaborate as a group to eventually reach a CIP
but early discussion (and some missed meetings since) have still left us at
square one to agree on what exactly the scope and content for a Cardano
oracle spec would look like.

--peter

**Comment by rphair** [comment]
Thanks @peterVG - without a concrete specification there seems little point developing a CIP around it & therefore have nothing to feed into the CIP review process... which is the whole point of submitting a CIP PR.  Likely by the time you finish your deliberations you will want to submit a different looking spec with different contextual documentation around it.

That, plus some intractable format problems with the original submission (e.g. using HTML tables rather than Markdown tables) suggest that this should be closed in favour of some completely different document & PR in the future.  If it ever looks otherwise please feel free to update this PR branch accordingly & reopen.  cc @codybutz

**Comment by peterVG** [comment]
My apologies for not finishing my reply with that point Robert. Yes, 100
percent yes that's where we are at. So yes, this issue needs to be closed
while we re-group. 🙏🏻
